### ISSUE TITLE
_id as ObjectId shouldn't throw error

### ISSUE THREAD

**Issue by gitawego** [Automattic/mongoose#4349]
in latest version (4.5.5), if _id is converted to ObjectId before query:
```js
    if ('_id' in $query) {
       $query._id = objectID($query._id);
    }
   Model.aggregate([{
     $match:$query
   }]).allowDiskUse(true).cursor({ batchSize: 100 }).exec();
```
mongoose throws error:
```
TypeError: Argument must be a string
    at TypeError (native)
    at Buffer.write (buffer.js:763:21)
    at serializeObjectId (path/to/node_modules/mongoose/node_modules/bson/lib/bson/parser/serializer.js:242:10)
    at serializeInto (path/to/node_modules/mongoose/node_modules/bson/lib/bson/parser/serializer.js:699:17)
```
I think this is too strict, if ``_id`` is already an ObjectId,  mongoose should simply stop casting it to ObjectId.

**Comment by vkarpov15** [comment]
What is `query._id` in your example? Mongoose doesn't do any casting on `.aggregate()` stages...

**Comment by faerulsalamun** [comment]
Maybe you use require('mongodb').ObjectID for convert to ObjectID ?

Because i have same problem after upgrading to node version 6.3.1

In my production server i use node 6.2.1 + mongoose 4.4.19 + require('mongodb').ObjectID still work

This work for me after change to mongoose.types.ObjectId("54505ced1fa5b1b519bdfc");

**Comment by gitawego** [comment]
@faerulsalamun yes indeed, I use ``require('mongodb').ObjectID`` to convert. gonna try with your method. thanks a lot

**Comment by gitawego** [comment]
yes, this is the problem, it seems ``require('mongodb').ObjectID`` is not compatible with ``mongoose.Types.ObjectId``, which is strange...

**Comment by vkarpov15** [comment]
Do you have a different version of mongodb? Can you show me the output of `npm list | grep "mongo"` in your repo? I suspect this is because you have a different version of the mongodb driver than mongoose uses

**Comment by faerulsalamun** [comment]
Output from my repo

```
├─┬ connect-mongo@1.3.1
│ └─┬ mongodb@2.2.4
│   ├─┬ mongodb-core@2.0.6
├─┬ mongoose@4.5.6
│ ├─┬ mongodb@2.1.18
│ │ └── mongodb-core@1.3.18
├─┬ mongoose-unique-validator@1.0.2
├─┬ mongoose-validator@1.2.5
```

**Comment by vkarpov15** [comment]
Yeah looks like npm is pulling mongodb from connect-mongo in your example because it's more recent. In general be very careful with `require()`-ing modules that you haven't explicitly imported in package.json, the problem is that mongodb 2.2.4 uses bson ~0.5, which is incompatible with mongodb 2.1.18 and bson ~0.4. Not something we're likely to fix unless there's lots of demand.

**Comment by devinus** [comment]
I'm experiencing this as well. `mongodb-extended-json` has also upgraded `bson` which we use to serialize/deserialize queries. :(

**Comment by vkarpov15** [comment]
Yeah mongoose 4.6.0 (just released today) uses bson 0.5, so that should fix your issue. Mongoose was delayed in upgrading because of a few minor issues with the mongodb driver detailed in #4363.
